Overhaul for IPv6 and flexiblity (#159)
cloudposse/terraform-aws-dynamic-subnets: docs/migration-v1-v2.md

## Migration Notes for Dynamic Subnets v2.0

The first version of `terraform-aws-dynamic-subnets` was written for Terraform v0.9.1, 

> which was so limited that it do not even have a boolean data type, and 

Reviewer: ```suggestion
which was so limited that it does not even have a boolean data type, and 
```

Once you have adapted to the new inputs, Terraform should be able to 
plan and apply with no substantive changes. 

#### Changes to Inputs

All the breaking changes are due to changes in inputs. Fortunately, adapting to the 
changes should be easy. In most cases, you simply change the name of the 
input and put the old value inside a list. The notable exception
is [`subnet_type_tag_key`](#subnet_type_tag_key-and-subnet_type_tag_value_format).

## Optional Inputs

Several inputs are optional. Previously, optional inputs were often
`string`s that could be empty or `null`. Unfortunately, due to Terraform [limitations](https://github.com/hashicorp/terraform/issues/26755#issuecomment-719103775),
we cannot condition the creation of a resource based on a value not known at 
plan time, which put severe limitations on when and how this module can be used.
To work around this limitation, any optional string input is now a `list(string)` that can have zero or one element, 
rather than a `string` that could be empty or `null`.

### `cidr_block` replaced with `ipv4_cidr_block`

Previously this module required an IPv4 CIDR block input as `cidr_block`.

> This value is now optional, and,  Since we had to make 

Reviewer: ```suggestion
This value is now optional, and, since we had to make 
```

> a breaking change in type, we took the opportunity to reduce the ambiguity

Reviewer: ```suggestion
a breaking change in its type, we took the opportunity to reduce the ambiguity
```

of `cidr_block` and renamed it `ipv4_cidr_block`.

***Migration***: Replace

```hcl
cidr_block = aws_vpc.main.cidr_block
```

with:

```hcl
ipv4_cidr_block = [aws_vpc.main.cidr_block]
```

### `igw_id`

Like `cidr_block`, the Internet Gateway ID `igw_id`
is now optional, and therefore was changed from `string` to `list(string)`.

***Migration***: Replace

```hcl
igw_id = aws_internet_gateway.default.id
```

with

```hcl
igw_id = [aws_internet_gateway.default.id]
```

Because the Internet Gateway ID is now optional, you

> can create a "public" set of subnets but not have the routed directly 

Reviewer: ```suggestion
can create a "public" set of subnets but not have them routed directly 
```

subnets in some way to distinguish them, we suggest that instead you look at [standard
tag keys](https://aws.amazon.com/premiumsupport/knowledge-center/eks-vpc-subnet-discovery/)
like `kubernetes.io/role/elb` and `kubernetes.io/role/internal-elb`
and add them to `public_subnets_additional_tags` and `private_subnets_additional_tags`
respectively.

### vpc_default_route_table_id -> public_route_table_ids (list)

Providing a route table ID for the public subnets was always optional,
but previously you could only supply a single route table ID to cover
all the public subnets. We have enhanced this feature to allow you to
supply separate route tables for each subnet, and so we have replaced
`vpc_default_route_table_id` with the more powerful `public_route_table_ids` variable.

***Migration***: Replace

```hcl
vpc_default_route_table_id = aws_vpc.main.default_route_table_id
```

with:

```hcl
public_route_table_ids = [aws_vpc.main.default_route_table_id]
```

## Removed variables

### `public_network_acl_id` and `private_network_acl_id`

The variables `public_network_acl_id` and `private_network_acl_id` have been removed.

> They did not work properly anyway, so it is unlikely you were using them.
> You can now control the creation of network ACLs via `public_open_network_acl_enabled`
> and `private_open_network_acl_enabled`.

Reviewer 2: ```suggestion
We discovered a bug in the code where supplying these variables did not behave as intended. The module now supports the creation of network ACLs via `public_open_network_acl_enabled` and `private_open_network_acl_enabled`.
```

Author: The module always supported the optional creation of network ACLs. What it failed to do was associate the provided ACL with the subnets.
